doc/openssl_migration_guide.rst

```diff
@@ -0,0 +1,389 @@
+generator for most applications.
+The ``randomize()`` function takes a pointer to the buffer and the size of the buffer as arguments.
+Botan provides a ``hex_encode()`` function that converts the random bytes to a hexadecimal string.
+
+Symmetric Encryption
+---------------------
+
+Consider the following application code to encrypt some data with AES using OpenSSL.
+
+.. code-block:: c
+
+    #include <openssl/aes.h>
+    #include <openssl/evp.h>
+    #include <iostream>
+    #include <iomanip>
+
+    int main() {
+        // Hex-encoded key and plaintext block
+        const char* key_hex = "000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F";
+        const char* plaintext_hex = "00112233445566778899AABBCCDDEEFF";
+
+        // Convert hex-encoded key and plaintext block to binary
+        unsigned char key[32], plaintext[16];
+        for(int i = 0; i < 32; i++) {
+            sscanf(&key_hex[i*2], "%02x", &key[i]);
+        }
+        for(int i = 0; i < 16; i++) {
+            sscanf(&plaintext_hex[i*2], "%02x", &plaintext[i]);
+        }
+
+        // Encrypt
+        unsigned char ciphertext[16];
+        AES_KEY encrypt_key;
+        AES_set_encrypt_key(key, 256, &encrypt_key);
+        AES_encrypt(plaintext, ciphertext, &encrypt_key);
+
+        // Decrypt
+        unsigned char decrypted[16];
+        AES_KEY decrypt_key;
+        AES_set_decrypt_key(key, 256, &decrypt_key);
+        AES_decrypt(ciphertext, decrypted, &decrypt_key);
+
+        // Print ciphertext and decrypted plaintext in hexadecimal format
+        for(int i = 0; i < 16; i++) {
+            printf("%02X", ciphertext[i]);
+        }
+        printf("\n");
+        for(int i = 0; i < 16; i++) {
+            printf("%02X", decrypted[i]);
+        }
+        printf("\n");
+
+        return 0;
+    }
+
+This example uses the ``AES_set_encrypt_key()`` and ``AES_encrypt()`` functions to encrypt
+a 128-bit plaintext block with a 256-bit key using AES. The key and plaintext block
+are hex-encoded and converted to binary before encryption.
+The ``AES_set_decrypt_key()`` and ``AES_decrypt()`` functions are used to decrypt the ciphertext.
+
+Here is the equivalent C++ code using Botan:
+
+.. literalinclude:: /../src/examples/aes.cpp
+   :language: cpp
+
+This example uses the ``AES_256`` cipher to encrypt a 128-bit plaintext block with a 256-bit
+key using AES. The key and plaintext block are hex-encoded and converted to binary before encryption.
+``set_key()`` is used to set the key and ``encrypt()`` is used to encrypt the plaintext block.
+Then, the ``AES_256`` object is even reused to encrypt a second plaintext with a different key.
+
+To learn more about the ``BlockCipher`` and ``CipherMode`` interfaces, including a list of all
+available block ciphers and cipher modes, see the :ref:`api_ref/block_cipher:block ciphers` and
+:ref:`api_ref/cipher_modes:cipher modes` handbook sections.
```

Comment:
I think we should rather showcase AES with a safe mode of operation (eg. CBC with padding -- we have an example for that). Maybe even better: an AEAD. Or at least note more prominently, that bare-metal usage of AES is not a very typical use case for end users.

Reply:
Good point, changed it to use the AES CBC example.